WIP: Documentation reorganization

melissawm opened the pull request:
Folks, this is a big PR but there's very little content actually changed. Most of the changes here are about the `_toc.yml` file and some re-wording to make the organization work.

This is very clearly a proposal and all feedback is welcome. There are a lot of placeholders which we don't have to keep if there are no actual documents, but I wanted to create the complete picture so that the TOC is clearer. 

One open question remaining is about the capitalization style for titles. I did remove the capitalization for some pages where I felt this would make things more consistent, but for a final version this needs to be decided. 

For context, see napari/napari#3390.

EDIT with current plan:
- [x] PRs to napari/napari and napari/napari.github.io with capitalization changes (napari/napari#3594 and #284)
- [x] re-org of toc and docs in napari/napari.github.io (#285, #299)
- [x] re-org of docs in napari/napari (napari/napari#3751)
- [ ] moving files between repos (napari/napari#3529) - PRs napari/napari#3815 and napari/napari.github.io#305

melissawm committed:
Documentation reorganization

See napari/napari#3390

GenevieveBuckley commented on _config.yml:
```diff
@@ -42,21 +42,21 @@ sphinx:
       - node_modules
       - theme/src
-    html_theme: furo
+    #html_theme: furo
```
If we;re replacing `furo` with the napari theme, then I'd probably remove this line entirely instead of just commenting it out.

melissawm replied:
Oh sure - I just don't think this is actually enabled right now in CI, but I could be wrong

codemonkey800 replied:
hi @melissawm, these configuration changes will actually need to be reverted to get the CI to work 😢 

this config + CI is what's needed to build the theme, but since you only have one enabled, it breaks the CI. we'll eventually enable the theme in this config + CI when it's ready 😄

GenevieveBuckley commented on _toc.yml:
```diff
@@ -1,42 +1,62 @@
+  - file: tutorials/start_index
+    sections: 
+    - file: tutorials/fundamentals/installation
+    - file: tutorials/absolute_beginners
```
What content is the absolute beginners page supposed to cover? (Apologies if this was already discussed elsewhere)

melissawm replied:
I was thinking of an absolute beginners page as a placeholder for the "Intro for non-Python users". There's no content for this yet, but I wanted to make sure this was in the TOC for this PR just so the entire context is captured.

sofroniewn replied:
sounds good

GenevieveBuckley commented on _toc.yml:
```diff
@@ -1,42 +1,62 @@
-      - file: tutorials/fundamentals/getting_started
-      - file: tutorials/fundamentals/viewer
-      - file: tutorials/fundamentals/image
-      - file: tutorials/fundamentals/labels
```
I think we somehow lost a lot of the fundamentals tutorial content from the TOC? Eg: fundamentals/image, labels, points, shapes, surface, tracks, vectors.

Edit: wait, no. It's all here, just moved into a 'layers' subsection. I'd guess that this hides the content a little more than we'd like - it seems easier to accidentally skip over now. New users might be likely to click on a too advanced topic because "layers" doesn't sound like a thing they need to know about (whereas "segmentation" does sound familiar)

DragaDoncila replied:
I think that's a valid concern @GenevieveBuckley, but I do see that logically they group nicely under that section. I know we're planning on having a "Layers at a glance" document introducing and comparing the different layer types. Maybe having that at top level in "Getting started" would help drive home the importance?

GenevieveBuckley replied:
I like both those ideas

melissawm replied:
That sounds reasonable, thanks!

melissawm replied:
I'm not really sure how to implement this right now. Initially i thought one good idea would be to include the Layers subdocuments twice, once in the Getting started section, and another in the How-tos section. However this causes the Layers document to be highlighted twice in the table of contents when the page is visited:

![Screenshot_20211004_150510](https://user-images.githubusercontent.com/3949932/135901693-7b4001ce-b4df-401b-b318-bfc58f1ff956.png)

I'll see if I can think of another solution.

melissawm replied:
I think I found a solution, see #268 .

melissawm replied:
Are you ok with the ideas in #268? If yes, I'll resolve this and start implementing those ideas there.

jni replied:
Just approved it!

GenevieveBuckley commented on _toc.yml:
```diff
@@ -1,42 +1,62 @@
+    - file: tutorials/applications/annotate_segmentation
+    - file: tutorials/applications/annotate_points
+    - file: tutorials/applications/dask
+    - file: tutorials/applications/cell_tracking
```
Suggestion: have separate subsections for "Segmentation" and "Tracking". Those are very common workflows and will likely hold many pages on different examples in the future.

melissawm replied:
If we do create these two subsections, do you propose we move the existing documents under there? For example:

**Tutorials:**
- Segmentation
   - Annotate segmentation
   - ...
- Tracking
   -  Cell tracking
   - ...
- Annotating videos with napari
- Using dask and napari
- Tracks layer tutorial

If we do this, note that the third level (documents inside the Segmentation section and Tracking section) will not appear in the main ToC. Is that what we want?

DragaDoncila replied:
I don't think it's a problem to just have the 'Segmentation' and 'Tracking' headers in the main TOC - users will very likely click through to these sections because they're such key applications

neuromusic replied:
Agreed w/ @DragaDoncila... the main topical sections are sufficient for the ToC I think

for reference, here's the set of categories we're planning to feature on the napari hub:

- Image registration
- Image correction
- Image enhancement
- Image reconstruction
- Pixel classification
- Image feature detection
- Image annotation
- _Filament tracing_
- Object classification
- _Object-based colocalisation_
- Object feature extraction
- Object tracking
- _Clustering_
- _Frequency domain analysis_
- _Pixel-based colocalisation_
- _Fluorescence correlation spectroscopy_
- _Optical flow analysis_
- Image Segmentation

Note: We do not expect the italicized categories to be populated with plugins at launch.

If we can keep the nomenclature consistent between napari.org and napari-hub.org, then we will have some good opportunities for cross-linking in the future

GenevieveBuckley commented on guides/stable/connecting_events.md:
```diff
@@ -1,6 +1,6 @@
 (connecting-events)=
 
-## Hooking up your own events
+# hooking up your own events
```
I personally would like to see us use sentence case for titles (eg: starting with a capital letter). This hasn't historically been the case, I'm not really sure why.

chili-chiu replied:
should this go with plugin section?

DragaDoncila replied:
> should this go with plugin section?

While this is commonly done in plugins, you can definitely hook up events in other workflows so I think this should live outside the plugin section.

GenevieveBuckley replied:
Maybe we need the concept of tags. 
There is a lot of content that we want, say in the plugin section, but also belongs somewhere else *as well*. If we can make these accessible from both places that might help

melissawm replied:
Yes, I do think tags would work great here as we don't have a huge amount of information yet and could create them rather easily.

GenevieveBuckley commented on tutorials/particle_trackpy.md:
```diff
@@ -0,0 +1,4 @@
+# tracks layer tutorial
+
+To be completed (see
+[issue #88](https://github.com/napari/napari.github.io/issues/88)).
```
I would really prefer we not put to-do notes in external, public facing documentation.

melissawm replied:
Right, it was more for identifying why this document is here, but I'll make sure to remove.

GenevieveBuckley commented:
Thanks for getting started on this @melissawm

melissawm commented:
I'll start addressing those comments. One thing you may have noticed is that I didn't actually move files around, so that would still need to be done once the general organization is agreed upon.

melissawm committed:
Applying sentence case to all docs.

melissawm committed:
Fix link and title for Developer Guides section

melissawm commented:
This is pretty much done from my side, as an initial reorganization. Ideally, the new theme can be merged first, and once this PR is merged I will actually need to update the documents in the main repo that are copied over here to reflect the changes in the deployed website.

neuromusic commented:
🎉  #201 has been merged 😁

melissawm committed:
Fixing _toc.yml indentation

melissawm commented:
If people are ok with the general idea, I'll implement these changes as necessary in this repo and napari/napari, but this PR will need to be closed. I'll create a plan for merging with minimal disruption for the site links and will post here.

GenevieveBuckley commented:
I just opened the [napari community resources page](https://napari.org/community/index.html) to find the zulip chat link for someone... and it turns out this info is kind of buried/inaccessible. 

Stuff I think should be obviously findable from the community page:
- the zulip chat link
- links to the other napari socials ([twitter](https://twitter.com/napari_imaging), [youtube channel](https://www.youtube.com/channel/UCbTgw84ew4pxTJ9qu3W2hqg/playlists), github)
- the napari meeting calendar (currently this is hidden inside the "About the Project and Team" page)

(If this is not the best place to add suggestions for your reorganization list, let me know what you prefer instead @melissawm)

DragaDoncila commented:
@GenevieveBuckley this isn't to say that we shouldn't make them more accessible within the pages too, but it's worth noting the zulip chat link, github, twitter and image.sc are all in the footer in the new theme

melissawm commented:
Will make sure to include those. For the record, next actions are to 
- Write a plan for how to merge this reorganization, as I will need to close this PR and split it into two: one for the parts that need to change on napari/napari.github.io, another for the parts in napari/napari. 
- Write a plan for where the content should live.

I will do that today, and share with the group in napari/napari#3390 . Once we agree, I can do the initial changes, *including documentation on where each document lives and how to make changes* so folks can contribute their own suggestions and refinements.

tlambert03 reviewed (approved):
this reorg looks good to me!  thanks @melissawm!

this makes me think we should add some lint tests to check for proper sentence casing in headers?  so someone can't undo all your work here 😅

neuromusic commented:
> this makes me think we should add some lint tests to check for proper sentence casing in headers?

could also force it client-side. but then headings of downloaded notebooks might not conform, so probably better to handle in CI

sofroniewn reviewed (approved):
LGTM! Thanks @melissawm. I like the new TOC cleans up somethings that were confusing.

I agree with @tlambert03 comment that we should add something to CI so that we don't loose all the capitalization fixes!

Maybe @jni can take a quick look at the TOC too and then @melissawm can press on with the next round of fixes.

It might also be worth doing a standalone PR to fix all the capitalization issues separate to the reorg as it will make reviewing easier with smaller diffs.

GenevieveBuckley commented:
> It might also be worth doing a standalone PR to fix all the capitalization issues separate to the reorg as it will make reviewing easier with smaller diffs.

Agreed. Could someone make an issue to track this? Thanks!

jni commented on _toc.yml:
```diff
@@ -1,43 +1,66 @@
+    - file: guides/stable/magicgui
+    - file: guides/stable/3D_interactivity
+    #- file: Glossary
+    title: In-depth guides
```
Does the ordering of the file/sections/title matter? It would be more intuitive for me for the title to come straight after the file, rather than at the end of the sections listing.

melissawm replied:
I'll have to check the jupyterbook docs, but I agree with you

jni commented:
@melissawm the TOC reorg makes sense, :+1: from me, but even knowing the [framework](https://diataxis.fr/) I get confused by the different section titles — someone unfamiliar with it would probably not be sure how "tutorials", "how-to's", and "guides" are different from one another. Can we brainstorm some better section titles? Some ideas:

- tutorials: getting started, basics, learn the basics
- how-tos: guides, use cases, cookbook, applications (or just how-to guides)
- guides: understanding napari, napari internals, under the hood

tlambert03 commented:
> @melissawm the TOC reorg makes sense, 👍 from me, but even knowing the [framework](https://diataxis.fr/) I get confused by the different section titles — someone unfamiliar with it would probably not be sure how "tutorials", "how-to's", and "guides" are different from one another.

I very much agree with this.  I know the Diátaxis framework has been considered the target for quite a while, _and_ I trust you all with this stuff more than I trust myself :joy: ... so I'm happy to keep going along with it to see where it leads (plus, I don't have any particularly useful suggestions as alternatives!).  But I've never felt a sense of "ah that's perfect!".  That is, i think I'll have to keep referring back to some definition of what the four sections actually mean.  

The [dask docs](https://docs.dask.org/en/stable/) (with getting started, fundamentals, and reference) do immediately give me a sense for what I'll find (basically, in increasing order of depth)...  the [trio docs](https://trio.readthedocs.io/en/stable/awesome-trio-libraries.html) have a similar layout (with tutorials followed by core functionality followed by design and internals).  those three sections are somewhat similar to @jni's terms:

- tutorials ~ getting started
- "how to's" is a tough one for me.  it feels too much like tutorials.  I like how dask uses "fundamentals" and trio uses "core functionality" for digging a bit deeper into the public-facing components  of the library
- "guides" 
  - a) i think this word overlaps in my mind a lot with "how to"... as in "guide me how to do something", right?
  - b) i think we should make it very clear whether we're guiding potential napari contributors, or napari users  (so I like how trio uses the term "a tour of internals" and "high level design principles".

basically, while this all may align just fine with the diataxis framework, I'm not entirely sure whether it does cause the wording has always confused me

jni commented:
> "how to's" is a tough one for me. it feels too much like tutorials.

To me there is a useful distinction here, though indeed it is difficult to state in 1-3 words. A tutorial teaches me how to use napari, period. It's very much a "getting started" doc, helps me to understand the basics of the API, how napari treats dimensions (displaying the last dims, sliders for the rest, for example) and how to use napari in general. A "how-to", in my head, I map to something like the [scikit-image gallery](https://scikit-image.org/docs/dev/auto_examples/), which doesn't really explain anything about "images are numpy arrays" or what filters are, but lets you at a glance find something close to your use case that you can copy-paste and modify at will.

jni commented:
More specifically: "how to guides" is where I would put an autogenerated gallery from https://github.com/napari/napari/tree/main/examples. These are generally not tutorials, but more like code templates.

GenevieveBuckley commented:
> basically, while this all may align just fine with the diataxis framework, I'm not entirely sure whether it does cause the wording has always confused me

yeah, I feel the same
(I think one major issue with it is that the framework suggests there are four discrete and non-overlapping categories, but that doesn't work so well especially for tutorials vs how-tos which I'd say are on the same spectrum)

If we need to emphasize the intended path a user should take through the documentation instead of what we think each bit corresponds to in the framework, in order to reduce confusion, I also support that.

jni commented:
Yup. In a DM conversation Talley and I discussed the framework and its shortcomings, two points that came out:

- learning/working and practical/theoretical are both continuous variables rather than boolean ones. So docs fall somewhere in that spectrum and pigeonholing them is often more difficult than the framework suggests.
- those are two important axes but not the only two. There's also beginner/expert in the concepts, and beginner/expert in Python, as just two extra examples. (Not to mention Python/no Python.)

DragaDoncila commented:
> learning/working and practical/theoretical are both continuous variables rather than boolean ones. So docs fall somewhere in that spectrum and pigeonholing them is often more difficult than the framework suggests.

I don't think the framework tries to set these up as "boolean" variables, but we certainly can't have an axis slider on our documentation organization 😄 . Ultimately, things will have to be pigeonholed somewhere.

> those are two important axes but not the only two. There's also beginner/expert in the concepts, and beginner/expert in Python, as just two extra examples. (Not to mention Python/no Python.)

Completely agree. However again, we can't store our documentation across an n-dimensional axis, unless we'd like to use the napari viewer to browse our documentation... 

Our plan for allowing users to easily browse across different dimensions of organization are tags. Tags such as "plugins" or "python beginner" can quickly collect for you all the relevant documentation if you don't want to browse by learning axes.

I think it's also important to note that any single lesson we want to teach could be any one or more of a how-to, a tutorial, an explanation or a reference. It's our job to decide which style of teaching we think is most pertinent for the user we have in mind. Let's take "preferences" as an example:
- I could write a "How to change the napari theme" how-to. 
- I could also write a "Learn to change your napari preferences" tutorial, that goes through more of the options and explains more about what each one does. 
- I could write a "Preference dialog under the hood" explanation that details how we've built it
- Finally, I could write a "Preference dialog reference" that just lists all the preferences and what their options are.

Any and all of these could be meaningful docs, depending on what our teaching goal is. So I don't think any lesson has a "right" document. Rather, we should consider who our lessons are for and what their purpose is, and tailor our document choices to that.

neuromusic commented:
Some context in how we approached using the diataxis framework for `starfish`...

The key use of starfish is to build an image processing workflow. We approached this with a user guide, which is a series of tutorials, each focused on a key step in building the workflow.

Tutorials were anchored in learning about implementing a given that step in the workflow, and guiding the user in rationalizing about the data, deciding which components to use, and why they would select certain parameters, etc. We provided data for each tutorial... it did not make sense for the user to bring their own. Our goal was to build a conceptual understanding of that step so that, when they had their own data, they could leverage the concepts that they learned in the tutorial. 

For example, here is the tutorial on normalizing intensity values: https://spacetx-starfish.readthedocs.io/en/latest/gallery/tutorials/normalize_intensity_values.html

For any given class that might be used in a tutorial, they often could do more things than what we presented in the tutorial. But those other uses were out of scope for the goal of the tutorial. In these cases, we made sure that there was a much much smaller "how to" focuses on how to use that specific feature and we would link out to them from the tutorials as necessary. The goal is to act as a reference for how to use the specific feature... something that can be easily referred to whenever this feature is being used. 

For example, here is the tutorial on the `ClipPercentileToZero` class, which is used in the Normalizing Intensities tutorial: https://spacetx-starfish.readthedocs.io/en/latest/gallery/how_to/clip_percentile_to_zero.html

I found the nomenclature of the diataxis framework to be incredibly useful in framing and building the documentation, in large part to be very clear with ourselves about what the goal was with every document we wrote. Are we trying to convey conceptual understanding? Does it make sense for a user to read this before even installing starfish? Or are we trying to help the user figure out how to use a specific feature?

By intentionally avoiding blurry "grey zone" and pushing ourselves toward the extremes of the spectrum, we were better able to make sure that the documentation met our user's needs, I think.

It may be worth noting, though, that we didn't end up organizing our docs according to the diataxis framework. Since the primary usage of starfish is somewhat linear (building a workflow), that was our guiding structure for our docs.

GenevieveBuckley commented:
> I found the nomenclature of the diataxis framework to be incredibly useful in framing and building the documentation, in large part to be very clear with ourselves about what the goal was with every document we wrote. Are we trying to convey conceptual understanding? Does it make sense for a user to read this before even installing starfish? Or are we trying to help the user figure out how to use a specific feature?
> ...
> It may be worth noting, though, that we didn't end up organizing our docs according to the diataxis framework. Since the primary usage of starfish is somewhat linear (building a workflow), that was our guiding structure for our docs.

Yes, I think this is what happens. The framework is good for clarifying what the goal is of every piece of documentation that is written (super, super useful to clarify this at the time of writing), but sometimes you need to stray from it as an organizational structure for the final location of docs.

(Also, I agree with what Talley says. I'm glad to hear someone else say that too, because I feel like that has been a very unpopular opinion whenever I've brought it up!)

melissawm commented:
Ok I was off for a few days and I am just catching up, let me try and join the conversation :)



> this makes me think we should add some lint tests to check for proper sentence casing in headers? so someone can't undo all your work here

I agree with @neuromusic this is probably best done in CI

> @melissawm the TOC reorg makes sense, +1 from me, but even knowing the [framework](https://diataxis.fr/) I get confused by the different section titles — someone unfamiliar with it would probably not be sure how "tutorials", "how-to's", and "guides" are different from one another. Can we brainstorm some better section titles? Some ideas:
> * tutorials: getting started, basics, learn the basics
> * how-tos: guides, use cases, cookbook, applications (or just how-to guides) 
> * guides: understanding napari, napari internals, under the hood

Agreed! The focus of the reorg is more on the high-level organization, the titles and content can definitely be changed if there are better ideas. I see @tlambert03's point about still not being able to locate certain documents and I'll try to think a little bit about this.

About the Diataxis framework: I see it as a general guide, but not a rulebook. There are and will be documents that overlap in categories, and it is likely that we'll need extra levels of organization such as the beginner/expert axis mentioned above. I think the distinction between tutorials and how-tos are an important one and I like @jni 's explanation. To add even more context, I think of tutorials as documents I'm looking for when I want something fairly complete, that tells a story from beginning to end and teaches me about concepts and why I should do things a certain way. It's certainly more useful for people still getting familiar with the subject. On the other hand, how-tos are the documents you keep getting back to even after being very familiar with the tool but you need a cookbook (I like this name!) with small tasks. It's not teaching you anything, it's just a reminder of the stuff you need to type to get a certain result.

In conclusion: I'm moving the PR into draft because I want to try and address the points that were raised here. This is great discussion, and tremendously helpful for me to understand what would be most useful for the napari community - feel free to elaborate on whatever points you feel are still vague. Also, just mentioning once more: I see the reorg as a flowing, moving thing. I think it's highly likely that refinements will come up after merging - once this is started it might give us all a clearer view, though.

melissawm commented:
Hello, folks! The biggest part of the reorg just landed! :tada: I'd like to ask all of you to take a look and give feedback if something isn't in the right place or you feel any of it could be improved. Like I've mentioned before, none of it is set in stone and there is definitely room for improvement if we feel something is not working. 

For this particular PR to be closed, there is still the file migration between repos (tutorial content) - I'll work on that next. However, there is valuable discussion here that could be captured somehow, in a roadmap or github discussion perhaps? If you have any suggestions, let me know.

Thanks again!

melissawm commented:
Following the discussion on the move between repos and how that might affect repo sizes, I think we can close this particular PR and investigate better solutions.

jni commented:
@melissawm I just want to be clear that I still favour the same approach as before (move all tutorials to main repo, using the same reorg we discussed), but removing the screenshots from the history, and hosting them on GitHub instead. See @DragaDoncila's [excellent "how-to" 😉](https://github.com/napari/napari/pull/3916#issuecomment-1010582291) for a reference on how to do it.
